File: test/fixtures/smelly.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"sync"
)

// Smell: Exported package-level variable
var GlobalCounter int = 0

// Smell: Ignored error return
func loadConfig() string {
	data, _ := ioutil.ReadFile("config.json") // BAD: error ignored
	return string(data)
}

// Smell: Missing context.Context for I/O operation
func fetchData(url string) ([]byte, error) {
	// This should accept context.Context as first parameter
	return nil, nil
}

// Smell: Mutex without defer unlock
func unsafeIncrement(mu *sync.Mutex, counter *int) {
	mu.Lock()
	*counter++
	// BAD: No defer mu.Unlock() - if panic occurs, mutex stays locked
	mu.Unlock()
}

// Smell: Empty interface abuse
func processAnything(data interface{}) interface{} {
	// BAD: Using interface{} loses type safety
	return data
}

// Smell: Goroutine leak potential
func startWorker() {
	go func() {
		for {
			// BAD: Infinite loop with no way to stop
			fmt.Println("working...")
		}
	}()
}

// Smell: Long function with deep nesting
func complexFunction(items []string) error {
	for _, item := range items {
		if item != "" {
			for i := 0; i < 10; i++ {
				if i%2 == 0 {
					switch item {
					case "a":
						if true {
							for j := 0; j < 5; j++ {
								// Deep nesting - hard to follow
								fmt.Println(item, i, j)
							}
						}
					case "b":
						fmt.Println("b")
					}
				}
			}
		}
	}
	return nil
}

// Smell: Missing error check
func writeFile() {
	f, err := os.Create("output.txt")
	// BAD: Using f before checking err
	f.WriteString("hello")
	if err != nil {
		return
	}
	f.Close() // Also BAD: ignoring close error
}

// Smell: Naked return in long function
func calculate(x, y int) (result int, err error) {
	if x < 0 {
		err = fmt.Errorf("x must be positive")
		return // Naked return - hard to follow what's being returned
	}

	if y < 0 {
		err = fmt.Errorf("y must be positive")
		return
	}

	result = x + y
	return
}

// Smell: Large struct passed by value
type HugeStruct struct {
	Data   [1024]byte
	Buffer [4096]byte
	Name   string
}

func processHuge(h HugeStruct) { // BAD: Should be *HugeStruct
	fmt.Println(h.Name)
}

// Smell: init() with complex logic
func init() {
	// BAD: Complex initialization that could fail
	data, err := ioutil.ReadFile("startup.conf")
	if err != nil {
		panic(err) // Panic in init is bad practice
	}
	fmt.Println(string(data))
}

func main() {
	GlobalCounter++
	loadConfig()
	startWorker()
}

File: test/fixtures/smelly.ts
/**
 * Intentionally bad code for testing code smell detection.
 */

// any abuse
export function processData(data: any) {
  const result: any = {};

  // Deep nesting
  if (data) {
    if (data.items) {
      if (data.items.length > 0) {
        for (let i = 0; i < data.items.length; i++) {
          if (data.items[i].active) {
            if (data.items[i].value > 0) {
              result[data.items[i].id] = data.items[i].value;
            }
          }
        }
      }
    }
  }

  return result;
}

// Bad naming
export const x = (a: any, b: any, c: any) => a + b + c;

export function fn(d: any) {
  let r = 0;
  for (let i = 0; i < d.length; i++) {
    r += d[i];
  }
  return r;
}

// God function - way too long
export function doEverything(input: any) {
  let result = 0;
  let temp = 0;
  let flag = false;
  let counter = 0;

  // Processing step 1
  if (input.type === "a") {
    result += 1;
    temp = result * 2;
    flag = true;
  }

  // Processing step 2
  if (input.type === "b") {
    result += 2;
    temp = result * 3;
    flag = false;
  }

  // Processing step 3
  for (let i = 0; i < 10; i++) {
    counter++;
    result += counter;
  }

  // Processing step 4
  if (flag) {
    result = result * temp;
  } else {
    result = result + temp;
  }

  // Processing step 5
  while (counter > 0) {
    result--;
    counter--;
  }

  // Processing step 6
  switch (input.mode) {
    case 1:
      result *= 2;
      break;
    case 2:
      result *= 3;
      break;
    case 3:
      result *= 4;
      break;
    default:
      result *= 1;
  }

  // Processing step 7
  if (result > 100) {
    result = 100;
  } else if (result < 0) {
    result = 0;
  }

  // More unnecessary processing
  temp = result;
  result = temp + 1;
  temp = result - 1;
  result = temp;

  return result;
}

// Missing error handling
export async function fetchData(url: string) {
  const response = await fetch(url);
  const data = await response.json();
  return data;
}

// Empty catch block
export function parseJSON(str: string) {
  try {
    return JSON.parse(str);
  } catch (e) {
    // Swallowed error - bad practice
  }
}

// Callback hell
export function callbackHell(callback: Function) {
  setTimeout(() => {
    fetch("/api/1").then((r) => {
      r.json().then((d) => {
        fetch("/api/2").then((r2) => {
          r2.json().then((d2) => {
            callback(d, d2);
          });
        });
      });
    });
  }, 1000);
}
